zipline/pipeline/loaders/events.py

```diff
@@ -0,0 +1,183 @@
+from six import iteritems
+from toolz import merge
+
+from .base import PipelineLoader
+from .frame import DataFrameLoader
+from .utils import next_date_frame, previous_date_frame, previous_value
+
+WRONG_COLS_ERROR = "Expected columns %s for sid %s but got columns %s."
+
+BAD_DATA_FORMAT_ERROR = ("Data for sid %s must be in DataFrame, "
+                         "Series, or DatetimeIndex.")
+
+SERIES_NO_DTINDEX_ERROR = ("Got Series for sid %d, but index was not "
+                           "DatetimeIndex.")
+
+DTINDEX_NOT_INFER_TS_ERROR = ("Got DatetimeIndex for sid %d.\n"
+                              "Pass `infer_timestamps=True` to use the first "
+                              "date in `all_dates` as implicit timestamp.")
+
+DF_NO_TS_NOT_INFER_TS_ERROR = ("Got DataFrame without a '%r' column for sid "
+                               "%d.\nPass `infer_timestamps=True` to use the "
+                               "first date in `all_dates` as implicit "
+                               "timestamp.")
+
+TS_FIELD_NAME = "timestamp"
+SID_FIELD_NAME = "sid"
+
+
+class EventsLoader(PipelineLoader):
+    """
+    Abstract loader.
+
+    Does not currently support adjustments to the dates of known events.
+
+    Parameters
+    ----------
+    all_dates : pd.DatetimeIndex
+        Index of dates for which we can serve queries.
+    events_by_sid : dict[int -> pd.DataFrame], dict[int -> pd.Series],
+    or dict[int -> pd.DatetimeIndex]
+        Dict mapping sids to objects representing dates on which earnings
+        occurred.
+
+        If a dict value is a Series, it's interpreted as a mapping from the
+        date on which we learned an announcement was coming to the date on
+        which the announcement was made.
+
+        If a dict value is a DatetimeIndex, it's interpreted as just containing
+        the dates that announcements were made, and we assume we knew about the
+        announcement on all prior dates.  This mode is only supported if
+        ``infer_timestamp`` is explicitly passed as a truthy value.
+        Dict mapping sids to DataFrames, Series, or DatetimeIndexes.
+
+        If the value is a DataFrame, it then represents dates on which events
+        occurred along with other associated values. If the DataFrame
+        contains a "timestamp" column, that column is interpreted as the date
+        on which we learned about the event. If the DataFrames do not contain a
+         "timestamp" column, we assume we knew about the event on all prior
+         dates.  This mode is only supported if ``infer_timestamp`` is
+         explicitly passed as a truthy value.
+
+    infer_timestamps : bool, optional
+        Whether to allow omitting the "timestamp" column.
+    dataset : DataSet
+        The DataSet object for which this loader loads data.
+    expected_cols : frozenset
+        Set of expected columns for the dataset, without timestamp.
+    """
+
+    @abc.abstractproperty
+    def expected_cols(self):
+        pass
```

Comment:
So, it makes sense to me why we'd have this, and I had a `NotImplemented` exception here originally, but it never gets there because it raises a `TypeError` first. Still necessary?